### PR TITLE
CHG: Connection#initialize now accepts a lambda for making socket

### PR DESCRIPTION
**Pull request by snow**
so we could inject some duck type of TCPSocket into Connection, e.g. Proxifier::Proxy wrapped socket.

### CODE REVIEW

File: lib/lowdown/connection.rb [hunk 1]
```diff
@@ -83,8 +83,25 @@ class TimedOut < StandardError; end
...
+    #          Proxifier::Proxy('http://127.0.0.1:3128').open \
+    #          uri.host, uri.port, nil, nil, Celluloid::IO::TCPSocket
+    #        end
+    #
+    #        connection_pool = Lowdown::Connection.pool \
+    #          size: 2,
+    #          args: [uri, cert.ssl_context, true, socket_maker]
+    #
+    #        client = Lowdown::Client.client_with_connection connection_pool, certificate: cert
+    #        ```
```
Comment (alloy) [inline lib/lowdown/connection.rb]: This should probably use a YARD `@example` block: http://www.rubydoc.info/gems/yard/file/docs/Tags.md#example

Comment (snow) [inline lib/lowdown/connection.rb]: sure. 10 minutes

File: lib/lowdown/client.rb [hunk 2]
```diff
@@ -223,7 +223,7 @@ def group(timeout: nil)
         end
       end
     ensure
-      group.terminate
+      group.try :terminate
```
Comment (alloy) [inline lib/lowdown/client.rb]: Why is this necessary?

**Comment by alloy** [comment]
`Object#try` is a Rails (ActiveSupport) feature, so please use standard `nil` checking and conditional dispatching instead.

**Comment by alloy** [comment]
Thanks, this looks like a good way to solve it 👍 

Can you please add a test? Once CI is green I’ll merge it.

**Comment by snow** [comment]
`group.try :terminate` is not related with the subject "proxy" of this PR. So I removed it.

I met `group` is nil on my production environment, didn't figure out why. That's a quick and dirty fix which accidentally included in this PR.

I will add test later.

File: lib/lowdown/connection.rb [hunk 3]
```diff
@@ -165,7 +182,7 @@ def connect!(tries = 0)
       # 2. This tries to `NilClass#send` the hostname:
       #    https://github.com/celluloid/celluloid-io/blob/85cee9da22ef5e94ba0abfd46454a2d56572aff4/lib/celluloid/io/dns_resolver.rb#L44
       begin
-        socket = TCPSocket.new(@uri.host, @uri.port)
+        socket = @socket_maker.try(:call, @uri) || TCPSocket.new(@uri.host, @uri.port)
```
Comment (snow) [inline lib/lowdown/connection.rb]: I will fix this `try` too.

**Comment by snow** [comment]
OK, test and README updated.

**Comment by alloy** [comment]
Ace! The build is showing 1 rubocop offence https://travis-ci.org/alloy/lowdown/builds/128844864

```
lib/lowdown/connection.rb:400:1: C: Trailing blank line missing.
```

Just add an extra newline at the end of that file and it should go green.

**Comment by snow** [comment]
done :)

**Comment by alloy** [comment]
Awesome, thanks 👏

**Comment by snow** [comment]
😄
